File: day14/day14.cpp
#include <fstream>
#include <iostream>
#include <numeric>
#include <vector>

#include <fmt/format.h>

using namespace std;

static const size_t SIZE = 256;
static const size_t HASH_SIZE = 16;

static void hashfn(const vector<size_t>& lengths, size_t rounds, unsigned char *out)
{
	for (size_t i = 0; i < SIZE; i++)
	{
		out[i] = i;
	}

	size_t cur = 0;
	size_t skip = 0;
	while (rounds-->0)
	{
		for(auto l: lengths)
		{
			// reverse from cur to l-1
			for (size_t i = 0; i < l/2; i++)
			{
				size_t a = (cur + i) % SIZE;
				size_t b = (cur + l - i - 1) % SIZE;
				swap(out[a], out[b]);
			}
			cur = (cur + l + skip) % SIZE;
			skip++;
		}
	}
}

static void knot(const string& line, unsigned char *out)
{
	vector<size_t> lengths(line.begin(), line.end());
	const size_t seq[] = {17, 31, 73, 47, 23};
	lengths.insert(lengths.end(), seq, seq+5);

	unsigned char buf[SIZE];
	hashfn(lengths, 64, buf);

	for (size_t i = 0; i < 256; i += 16)
	{
		unsigned char v = buf[i];
		for (size_t j = i+1; j < i+16; j++)
		{
			v ^= buf[j];
		}
		*out++ = v;
	}
}

static vector<string> make_map(const string& salt)
{
	vector<string> map;
	for (int i = 0; i < 128; i++)
	{
		unsigned char buf[HASH_SIZE];
		knot(fmt::format("{}-{}", salt, i), buf);

		string s;
		for (size_t j = 0; j < HASH_SIZE; j++)
		{
			for (int k = 7; k >= 0; k--)
			{
				s.push_back((buf[j] & (1U<<k)) ? '#' : '.');
			}
		}
		map.emplace_back(move(s));
	}
	return map;
}

static size_t part1(const vector<string>& map)
{
	return accumulate(
		map.begin(),
		map.end(),
		0ULL,
		[](auto acc, const auto& row) {
			return acc + count(row.begin(), row.end(), '#');
		});
}

static void dfs(const vector<string>&m, int x, int y, vector<vector<bool>>& visited)
{
	static const int dx[] = {0, 1, 0, -1};
	static const int dy[] = {1, 0, -1, 0};
	visited[y][x] = true;
	for (size_t i = 0; i < 4; i++)
	{
		int nx = x + dx[i];
		int ny = y + dy[i];
		if (0 <= nx && nx < 128 && 0 <= ny && ny < 128
		    && m[ny][nx] == '#' && !visited[ny][nx])
		{
			dfs(m, nx, ny, visited);
		}
	}
}

static size_t part2(const vector<string>& m)
{
	vector<vector<bool>> visited(m.size());
	for (auto &v: visited)
	{
		v.resize(m.size());
	}

	size_t groups = 0;
	for (int y = 0; y < 128; y++)
	{
		for (int x = 0; x < 128; x++)
		{
			if (m[y][x] == '#' && !visited[y][x])
			{
				dfs(m, x, y, visited);
				groups++;
			}
		}
	}
	return groups;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	string salt;
	getline(input, salt);
	input.close();
	if (input.fail())
	{
		cerr << "Cannot parse the data" << endl;
		return 1;
	}

	auto m = make_map(salt);
	cout << fmt::format("Part1: {}", part1(m)) << endl
	     << fmt::format("Part2: {}", part2(m)) << endl;
	return 0;
}

File: day14/day14.py
#!/usr/bin/env python3

SIZE = 256

def hash(lengths, rounds):
    lst = [i for i in range(SIZE)]

    cur = 0
    skip = 0
    for j in range(rounds):
        for l in lengths:
            # reverse from cur to l-1
            for i in range(l//2):
                a = (cur + i) % SIZE
                b = (cur + l - 1 - i) % SIZE
                lst[a], lst[b] = lst[b], lst[a]

            cur = (cur + l + skip) % SIZE
            skip = (skip + 1) % SIZE

    return lst

def knot(txt):
    l = [ord(i) for i in txt]
    l.extend([17,31,73,47,23])
    h = hash(l, 64)
    lst = []
    for i in range(0, SIZE, 16):
        v = h[i]
        for j in range(i+1, i+16):
            v ^= h[j]
        lst.append("{:02x}".format(v))
    return "".join(lst)

def make_map(txt):
    m = []
    hexs = "0123456789abcdef"
    bits = 0
    for i in range(128):
        row = []
        for c in knot("{}-{}".format(txt, i)):
            i = hexs.find(c)
            for j in range(4):
                if i & 1<<3:
                    row.append("#")
                else:
                    row.append(".")
                i <<= 1
        m.append("".join(row))
    return m

def part1(m):
    return sum(row.count("#") for row in m)

def dfs(m, x, y, visited):
    visited[y][x] = True
    for dx, dy in ((0,-1), (1, 0), (0, 1), (-1, 0)):
        nx = x + dx
        ny = y + dy
        if nx < 0 or nx >= 128 or ny < 0 or ny >= 128:
            continue
        if m[y][x] != "#" or visited[ny][nx]:
            continue
        dfs(m, nx, ny, visited)

def part2(m):
    visited = [[False for v in r] * 128 for r in m]
    groups = 0
    for y, row in enumerate(m):
        for x, val in enumerate(row):
            if val == "#" and not visited[y][x]:
                dfs(m, x, y, visited)
                groups += 1
    return groups

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    m = make_map(txt)
    for i in range(8):
        print(m[i][:8])
    print("Part1:", part1(m))
    print("Part2:", part2(m))
